### PR TITLE
[WIP] Add complete_mode()

### PR DESCRIPTION
**Pull request by Shougo**
I have added `complete_mode()` API for deoplete.

Note: I will send the PR for Vim8 later.  ~~So unit tests are not added now.~~

Usage: https://github.com/Shougo/deoplete.nvim/pull/920

### CODE REVIEW

**Comment by Shougo** [comment]
I have sent it to vim-dev.
https://github.com/vim/vim/pull/3866

File: src/nvim/edit.c [hunk 1]
```diff
@@ -2486,9 +2486,9 @@ void set_completion(colnr_T startcol, list_T *list)
 
 
 // complete_mode() implementation.
-char *ins_compl_mode(void)
+const char *ins_compl_mode(void)
 {
-  char *mode = NULL;
+  char_u *mode = NULL;
```
Comment (bfredl) [inline src/nvim/edit.c]: In nvim it is simpler to use `char *` here and cast only once  in `f_complete_mode`.

File: src/nvim/edit.c [hunk 2]
```diff
@@ -2485,6 +2485,64 @@ void set_completion(colnr_T startcol, list_T *list)
 }
 
 
+// complete_mode() implementation.
+const char *ins_compl_mode(void)
```
Comment (bfredl) [inline src/nvim/edit.c]: `const char *` would imply caller shouldn't free. But we could keep `const char*` here and remove all `xstrdup` and only have one `xstrdup` call in `f_complete_mode` instead.

Comment (Shougo) [reply src/nvim/edit.c]: OK.  I have fixed it.

**Comment by prabirshrestha** [comment]
@Shougo would it be possible to update this patch to reflect the latest `complete_info` changes. Tried the vim patch with asyncomplete v2 branch and seems to work. https://github.com/prabirshrestha/asyncomplete.vim/pull/124#issuecomment-471858892

**Comment by Shougo** [comment]
The patch will be updated if Vim's patch is merged.

**Comment by prabirshrestha** [comment]
This is now merged in vim. Can we get the PR updated? https://github.com/vim/vim/commit/fd133323d4e1cc9c0e61c0ce357df4d36ea148e3

**Comment by Shougo** [comment]
https://github.com/neovim/neovim/pull/9815

Done.

**Pull request closed without merging**